### PR TITLE
feat: add `--diff` flag to `archive export`

### LINKED ISSUE #3204: `forest-cli archive export --epoch=X --diff=Y`
Full snapshots contain at least 900 state trees and every block header in the chain. Snapshots are sets of IPLD values, and diffs can be computed using regular set difference.

Tasks:
- [ ] If `--diff=Y` is given as an argument to `forest-cli archive export`, exclude anything reachable from epoch `Y` from the exported snapshot. `Y` must be less than the exported epoch.
- [ ] Use `forest-cli snapshot verify` to verify that that a snapshot at epoch Y + diff from Y to X works just as a snapshot at epoch X. This is blocked by `forest-cli snapshot verify` supporting loading multiple CAR files at once.
- [ ] Test snapshot diffing in CI.

### PR DESCRIPTION
**Pull request by lemmih**
## Summary of changes

Changes introduced in this pull request:

- add `--diff` flag for exporting the set difference of two snapshots.

## Reference issue to close (if applicable)

Closes #3204

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: src/ipld/util.rs [hunk 1]
```diff
@@ -293,13 +304,28 @@ pub fn stream_chain<DB: Blockstore, T: Iterator<Item = Tipset> + Unpin>(
...
         tipset_iter,
         db,
         dfs: VecDeque::new(),
         seen: CidHashSet::default(),
         stateroot_limit,
+        fail_on_dead_links: true,
+    }
+}
+
+pub fn stream_graph<DB: Blockstore, T: Iterator<Item = Tipset> + Unpin>(
```
Comment (hanabi1224) [inline src/ipld/util.rs:318]: Any way to indicate `fail_on_dead_links: false` in the function name or doc? e.g. `stream_graph_unstrict`

File: src/cli/subcommands/archive_cmd.rs [hunk 2]
```diff
@@ -69,6 +72,9 @@ pub enum ArchiveCommands {
         /// How many state-roots to include. Lower limit is 900 for `calibnet` and `mainnet`.
         #[arg(short, long, default_value_t = 2000)]
         depth: ChainEpochDelta,
+        /// Do not include any values reachable from this epoch
+        #[arg(short, long)]
+        diff: Option<ChainEpochDelta>,
```
Comment (hanabi1224) [inline src/cli/subcommands/archive_cmd.rs:77]: It's a bit confusing, doc says epoch while code says epoch delta

File: src/cli/subcommands/archive_cmd.rs [hunk 3]
```diff
@@ -165,6 +173,18 @@ async fn do_export<ReaderT: Read + Seek + Send + Sync>(
         .tipset_by_height(epoch, ts, ResolveNullTipset::TakeOlder)
         .context("unable to get a tipset at given height")?;
 
+    let seen = if let Some(diff) = diff {
+        let diff_ts: Arc<Tipset> = index
+            .tipset_by_height(diff, ts.clone(), ResolveNullTipset::TakeOlder)
+            .context("diff epoch must be smaller than target epoch")?;
+        let diff_ts: &Tipset = &diff_ts;
+        let mut stream = stream_graph(&store, diff_ts.clone().chain(&store));
+        while stream.try_next().await?.is_some() {}
```
Comment (hanabi1224) [inline src/cli/subcommands/archive_cmd.rs:182]: Does it make sense to warn in `into_seen` when the stream is not fully consumed or  move stream consuming logic into `into_seen`?

**Review by hanabi1224 — APPROVED** [review]

File: src/cli/subcommands/archive_cmd.rs (hunk 2 above)
Comment (lemmih) [reply src/cli/subcommands/archive_cmd.rs:77]: Updated to `epoch-diff`.

File: src/ipld/util.rs (hunk 1 above)
Comment (lemmih) [reply src/ipld/util.rs:321]: I've updated the documentation of both `stream_graph` and `stream_chain`.

File: src/cli/subcommands/archive_cmd.rs (hunk 3 above)
Comment (lemmih) [reply src/cli/subcommands/archive_cmd.rs:182]: You don't always want to consume the entire stream.

**Review by LesnyRumcajs — APPROVED** [review]

**Pull request merged**